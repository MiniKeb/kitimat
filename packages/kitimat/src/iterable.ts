if (!Symbol.asyncIterator) {
  (Symbol as any).asyncIterator = Symbol.for('Symbol.asyncIterator');
}

const getIterator = <A>(iter: AsyncIterable<A>): AsyncIterator<A> => iter[Symbol.asyncIterator]();

export const create = <A>(fn: () => AsyncIterator<A>): AsyncIterable<A> => ({
  [Symbol.asyncIterator]: fn,
});

export const toArray = async <A>(iter: AsyncIterable<A>): Promise<A[]> => {
  const result = [];

  for await (let value of iter) {
    result.push(value);
  }

  return result;
};

export const empty = <A>(): AsyncIterable<A> => create<A>(async function*(): AsyncIterator<A> {});

export const from = <A>(arr: (A | Promise<A>)[]): AsyncIterable<A> =>
  create<A>(async function*() {
    yield* arr;
  });

export const of = <A>(a: A | Promise<A>): AsyncIterable<A> =>
  create<A>(async function*(): AsyncIterator<A> {
    yield await a;
  });

export const every = async <A>(fn: (a: A) => boolean | Promise<boolean>, iter: AsyncIterable<A>): Promise<boolean> => {
  for await (let next of iter) {
    if ((await fn(next)) === false) {
      return false;
    }
  }

  return true;
};

export const some = async <A>(fn: (a: A) => boolean | Promise<boolean>, iter: AsyncIterable<A>): Promise<boolean> => {
  for await (let next of iter) {
    if ((await fn(next)) === true) {
      return true;
    }
  }

  return false;
};

export const lazy = <A>(fn: () => AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    yield* fn();
  });

export const map = <A, B>(fn: (a: A) => B | Promise<B>, iter: AsyncIterable<A>): AsyncIterable<B> =>
  create<B>(async function*() {
    for await (let next of iter) {
      yield await fn(next);
    }
  });

export const map2 = <A, B, C>(
  fn: (a: A, b: B) => C | Promise<C>,
  a: AsyncIterable<A>,
  b: AsyncIterable<B>,
): AsyncIterable<C> =>
  create<C>(async function*() {
    const iteratorA = getIterator(a);
    const iteratorB = getIterator(b);

    let resultA: IteratorResult<A>;
    let resultB: IteratorResult<B>;

    do {
      // do this so that they resolve themselves in parallel
      [resultA, resultB] = [await iteratorA.next(), await iteratorB.next()];

      if (resultA.done === false && resultB.done === false) {
        yield await fn(resultA.value, resultB.value);
      }
    } while (resultA.done === false && resultB.done === false);
  });

export const zip = <A, B>(a: AsyncIterable<A>, b: AsyncIterable<B>): AsyncIterable<[A, B]> =>
  map2<A, B, [A, B]>((a_, b_) => [a_, b_], a, b);

export const zip3 = <A, B, C>(
  a: AsyncIterable<A>,
  b: AsyncIterable<B>,
  c: AsyncIterable<C>,
): AsyncIterable<[A, B, C]> => map2<[A, B], C, [A, B, C]>(([a_, b_], c_) => [a_, b_, c_], zip(a, b), c);

export const zip4 = <A, B, C, D>(
  a: AsyncIterable<A>,
  b: AsyncIterable<B>,
  c: AsyncIterable<C>,
  d: AsyncIterable<D>,
): AsyncIterable<[A, B, C, D]> =>
  map2<[A, B], [C, D], [A, B, C, D]>(([a_, b_], [c_, d_]) => [a_, b_, c_, d_], zip(a, b), zip(c, d));

export const filter = <A>(fn: (a: A) => Promise<boolean> | boolean, iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    for await (let next of iter) {
      if ((await fn(next)) === true) {
        yield next;
      }
    }
  });

export const scan = <A>(fn: (a: A) => Promise<A> | A, init: A): AsyncIterable<A> =>
  create<A>(async function*() {
    let next = init;
    yield next;

    while (true) {
      next = await fn(next);
      yield next;
    }
  });

export const reduce = async <A, B>(
  fn: (prev: A, next: B) => A | Promise<A>,
  acc: A,
  iter: AsyncIterable<B>,
): Promise<A> => {
  let prev = acc;

  for await (let next of iter) {
    prev = await fn(prev, next);
  }

  return prev;
};

export const take = <A>(i: number, iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    let j = 0;

    for await (let next of iter) {
      yield next;

      if (++j >= i) {
        break;
      }
    }
  });

export const drop = <A>(i: number, iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    let j = 0;

    for await (let next of iter) {
      if (++j <= i) {
        continue;
      }

      yield next;
    }
  });

export const range = (low: number, high: number): AsyncIterable<number> => {
  if (low > high) {
    return range(high, low);
  }

  return create<number>(async function*() {
    for (let i = low; i < high; i++) {
      yield i;
    }
  });
};

export const cons = <A>(a: A | Promise<A>, xs: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    yield await a;

    for await (let next of xs) {
      yield next;
    }
  });

export const concat = <A>(a: AsyncIterable<A>, b: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    for await (let next of a) {
      yield next;
    }

    for await (let next of b) {
      yield next;
    }
  });

export const concat3 = <A>(a: AsyncIterable<A>, b: AsyncIterable<A>, c: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    yield* concat(concat(a, b), c);
  });

export const concat4 = <A>(
  a: AsyncIterable<A>,
  b: AsyncIterable<A>,
  c: AsyncIterable<A>,
  d: AsyncIterable<A>,
): AsyncIterable<A> =>
  create<A>(async function*() {
    yield* concat(concat3(a, b, c), d);
  });

export const flatMap = <A, B>(fn: (a: A) => AsyncIterable<B>, iter: AsyncIterable<A>): AsyncIterable<B> =>
  create<B>(async function*() {
    for await (let xs of map(fn, iter)) {
      yield* xs;
    }
  });

export const endless = <A>(iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    let isEmpty = true;

    while (true) {
      for await (let value of iter) {
        isEmpty = false;
        yield value;
      }

      if (isEmpty) {
        break;
      }
    }
  });

export const always = <A>(a: A): AsyncIterable<A> =>
  create<A>(async function*() {
    while (true) {
      yield a;
    }
  });

export const reverse = <A>(iter: AsyncIterable<A>): AsyncIterable<A> =>
  create<A>(async function*() {
    const cache = [];

    for await (let value of iter) {
      cache.unshift(value);
    }

    yield* cache;
  });

export const cached = <A>(iter: AsyncIterable<A>): AsyncIterable<A> => {
  const cache: A[] = [];
  const iterator = getIterator(iter);
  let done = false;

  return create<A>(async function*() {
    yield* cache;

    if (done === true) {
      return;
    }

    let next: IteratorResult<A>;

    do {
      next = await iterator.next();

      if (next.done === false) {
        const value = next.value;
        cache.push(value);
        yield value;
      }
    } while (next.done === false);

    done = true;
  });
};
