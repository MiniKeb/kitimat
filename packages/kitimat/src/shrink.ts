import * as Iter from './iterable';
import * as Util from './utilities';

export type Shrinker<A> = (a: A) => AsyncIterable<A>;

const convert = <A, B>(bToA: (b: B) => A, aToB: (a: A) => B, shrinker: Shrinker<A>): Shrinker<B> => (b: B) =>
  Iter.map(i => aToB(i), shrinker(bToA(b)));

export const boolean: Shrinker<boolean> = n => {
  if (n === true) {
    return Iter.of(false);
  }

  return Iter.empty();
};

const seriesInt = (low: number, high: number): AsyncIterable<number> =>
  Iter.create<number>(async function*() {
    if (low >= high) {
      return;
    }

    yield low;

    if (low === high - 1) {
      return;
    }

    const next = low + Math.round((high - low) / 2);
    yield* seriesInt(next, high);
  });

const seriesFloat = (low: number, high: number): AsyncIterable<number> =>
  Iter.create<number>(async function*() {
    if (low >= high - 0.0001) {
      if (high !== 0.000001) {
        yield low;
      }

      return;
    }

    yield low;

    const next = low + (high - low) / 2;
    yield* seriesFloat(next, high);
  });

type SeriesFunction = (low: number, high: number) => AsyncIterable<number>;

const numberHelper = (makeSeries: SeriesFunction) => (pivot: number): Shrinker<number> => n => {
  if (pivot === 0 && n === 0) {
    return Iter.empty();
  }

  // go toward pivot from negative but end at zero
  if (pivot >= 0 && n < 0) {
    return Iter.map(m => -1 * m, makeSeries(0, -n));
  }

  // go toward pivot from positive but end at zero
  if (pivot < 0 && n >= 0) {
    return makeSeries(0, n);
  }

  // both pivot and n are positive, so n goes toward n
  if (pivot >= 0 && n >= 0) {
    if (n > pivot) {
      // pivot is less than n, so just start with pivot and go to n
      return makeSeries(pivot, n);
    } else {
      // pivot is greater than n, make them both negative so that
      // negative pivot is less than negative n (so that we start with
      // the pivot). make series, then map all by -1.
      return Iter.map(m => -1 * m, makeSeries(-pivot, -n));
    }
  }

  if (pivot < 0 && n < 0) {
    if (n > pivot) {
      return makeSeries(pivot, n);
    } else {
      return Iter.map(m => -1 * m, makeSeries(-pivot, -n));
    }
  }

  // impossible
  return Iter.empty();
};

export const atLeastInteger = numberHelper(seriesInt);
export const atLeastFloat = numberHelper(seriesFloat);

export const atLeastCharacter = (int: number): Shrinker<string> =>
  convert<number, string>((str: string) => str.charCodeAt(0), String.fromCharCode, atLeastInteger(int));

export const character: Shrinker<string> = atLeastCharacter(32);

export const array = <A>(shrinker: Shrinker<A>): Shrinker<A[]> => {
  const shrinkOne = (l: A[]): AsyncIterable<A[]> => {
    if (l.length === 0) {
      return Iter.empty();
    }

    const [head, ...tail] = l;
    return Iter.concat(Iter.map(a => [a, ...tail], shrinker(head)), Iter.map(arr => [head, ...arr], shrinkOne(tail)));
  };

  const removes = <B>(size: number, arrayLength: number, arr: B[]): AsyncIterable<B[]> => {
    if (size > arrayLength) {
      return Iter.empty();
    } else {
      const first = Util.take(size, arr);
      const rest = Util.drop(size, arr);
      const recursed = removes(size, arrayLength - size, rest);
      const appended = Iter.map<B[], B[]>(xs => [...xs, ...first], recursed);
      return Iter.cons(rest, appended);
    }
  };

  const makeRemovable = <B>(a: B[]): AsyncIterable<B[]> =>
    Iter.create<B[]>(async function*() {
      let len = a.length;
      const initLen = len;
      while (len > 0) {
        yield* removes(len, initLen, a);
        len = Math.floor(len / 2);
      }
    });

  return (a: A[]): AsyncIterable<A[]> => {
    const removable = makeRemovable(a);
    const shrunkOne = shrinkOne(a);
    return Iter.concat(removable, shrunkOne);
  };
};

export const string = convert<string[], string>(
  (str: string) => str.split(''),
  (arr: string[]) => arr.join(''),
  array(character),
);
