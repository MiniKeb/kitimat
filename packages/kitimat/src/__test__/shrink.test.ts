import * as Iter from '../iterable';
import * as Shrink from '../shrink';

const lessThanVal = (n: number) => (m: number) => {
  expect(m).toBeLessThan(n);
};

const greaterThanVal = (n: number) => (m: number) => {
  expect(m).toBeGreaterThan(n);
};

test('boolean', async () => {
  const sTrue = await Iter.toArray(Shrink.boolean(true));
  const sFalse = await Iter.toArray(Shrink.boolean(false));

  expect(sTrue).toEqual([false]);
  expect(sFalse).toEqual([]);
});

test('integer', async () => {
  const s20 = await Iter.toArray(Shrink.atLeastInteger(0)(20));
  const sNeg10 = await Iter.toArray(Shrink.atLeastInteger(0)(-10));
  const s5000 = await Iter.toArray(Shrink.atLeastInteger(0)(5e3));
  const s0 = await Iter.toArray(Shrink.atLeastInteger(0)(0));

  expect(s20[0]).toEqual(0);
  expect(sNeg10[0]).toEqual(-0);
  expect(s5000[0]).toEqual(0);
  expect(s0).toEqual([]);

  s20.forEach(lessThanVal(20));
  sNeg10.forEach(greaterThanVal(-10));
  s5000.forEach(lessThanVal(5000));

  expect(s20).toMatchSnapshot();
  expect(sNeg10).toMatchSnapshot();
  expect(s5000).toMatchSnapshot();
});

test('atLeastInteger 2', async () => {
  const bet5and20 = await Iter.toArray(Shrink.atLeastInteger(5)(20));
  const bet1and10 = await Iter.toArray(Shrink.atLeastInteger(1)(10));
  const bet9and10 = await Iter.toArray(Shrink.atLeastInteger(9)(10));
  const betNeg5andNeg10 = await Iter.toArray(Shrink.atLeastInteger(-5)(-10));
  const bet5andNeg20 = await Iter.toArray(Shrink.atLeastInteger(5)(-20));

  expect(bet5and20.length).toBeGreaterThan(0);
  bet5and20.forEach(greaterThanVal(4));
  bet5and20.forEach(lessThanVal(20));

  expect(bet1and10.length).toBeGreaterThan(0);
  bet1and10.forEach(greaterThanVal(0));
  bet1and10.forEach(lessThanVal(10));

  expect(bet9and10.length).toBeGreaterThan(0);
  bet9and10.forEach(greaterThanVal(8));
  bet9and10.forEach(lessThanVal(10));

  expect(betNeg5andNeg10.length).toBeGreaterThan(0);
  betNeg5andNeg10.forEach(greaterThanVal(-11));
  betNeg5andNeg10.forEach(lessThanVal(-4));

  expect(bet5andNeg20.length).toBeGreaterThan(0);
  bet5andNeg20.forEach(greaterThanVal(-21));
  bet5andNeg20.forEach(lessThanVal(5));

  expect(bet5and20).toMatchSnapshot();
  expect(bet1and10).toMatchSnapshot();
  expect(bet9and10).toMatchSnapshot();
  expect(betNeg5andNeg10).toMatchSnapshot();
  expect(bet5andNeg20).toMatchSnapshot();
});

test('float', async () => {
  const sA = await Iter.toArray(Shrink.atLeastFloat(0)(20.5));
  const sB = await Iter.toArray(Shrink.atLeastFloat(0)(-10.1));

  expect(sA[0]).toEqual(0);
  expect(sB[0]).toEqual(-0);

  sA.forEach(lessThanVal(20.5));
  sB.forEach(greaterThanVal(-10.1));

  expect(sA).toMatchSnapshot();
  expect(sB).toMatchSnapshot();
});

test('atLeastFloat 2', async () => {
  expect(await Iter.toArray(Shrink.atLeastFloat(10.1)(20.5))).toMatchSnapshot();
  expect(await Iter.toArray(Shrink.atLeastFloat(-5)(-10))).toMatchSnapshot();
  expect(await Iter.toArray(Shrink.atLeastFloat(5)(-20))).toMatchSnapshot();
  expect(await Iter.toArray(Shrink.atLeastFloat(-1)(5))).toMatchSnapshot();
});

test('char/atLeastChar/character', async () => {
  expect(await Iter.toArray(Shrink.character('Z'))).toMatchSnapshot();
});

test('array', async () => {
  const arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];
  const sArr = await Iter.toArray(Shrink.array(Shrink.atLeastInteger(0))(arr));

  sArr.forEach((a: number[]) => {
    expect(Array.isArray(a)).toEqual(true);
    expect(a.length).toBeLessThanOrEqual(arr.length);
  });

  expect(sArr.some((a: number[]) => a.length < arr.length)).toEqual(true);
  expect(sArr[0]).toEqual([]);
  expect(sArr).toMatchSnapshot();
});

test('string', async () => {
  const str = 'gabe1234567890qq';
  const sStr = await Iter.toArray(Shrink.string(str));

  sStr.forEach((s: string) => {
    expect(s.length).toBeLessThanOrEqual(str.length);
  });

  expect(sStr.some((s: string) => s.length < str.length)).toEqual(true);
  expect(sStr[0]).toEqual('');
  expect(sStr).toMatchSnapshot();
});
