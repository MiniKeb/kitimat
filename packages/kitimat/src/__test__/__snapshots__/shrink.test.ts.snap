// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array 1`] = `
Array [
  Array [],
  Array [
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
  ],
  Array [
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    7,
    8,
    9,
    10,
    11,
    12,
    1,
    2,
    3,
  ],
  Array [
    10,
    11,
    12,
    4,
    5,
    6,
    1,
    2,
    3,
  ],
  Array [
    7,
    8,
    9,
    4,
    5,
    6,
    1,
    2,
    3,
  ],
  Array [
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    1,
  ],
  Array [
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    2,
    1,
  ],
  Array [
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    3,
    2,
    1,
  ],
  Array [
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    4,
    3,
    2,
    1,
  ],
  Array [
    7,
    8,
    9,
    10,
    11,
    12,
    5,
    4,
    3,
    2,
    1,
  ],
  Array [
    8,
    9,
    10,
    11,
    12,
    6,
    5,
    4,
    3,
    2,
    1,
  ],
  Array [
    9,
    10,
    11,
    12,
    7,
    6,
    5,
    4,
    3,
    2,
    1,
  ],
  Array [
    10,
    11,
    12,
    8,
    7,
    6,
    5,
    4,
    3,
    2,
    1,
  ],
  Array [
    11,
    12,
    9,
    8,
    7,
    6,
    5,
    4,
    3,
    2,
    1,
  ],
  Array [
    12,
    10,
    9,
    8,
    7,
    6,
    5,
    4,
    3,
    2,
    1,
  ],
  Array [
    11,
    10,
    9,
    8,
    7,
    6,
    5,
    4,
    3,
    2,
    1,
  ],
  Array [
    0,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    0,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    1,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    0,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    2,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    0,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    2,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    3,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    0,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    3,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    4,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    0,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    3,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    5,
    7,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    0,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    4,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    6,
    8,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    0,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    4,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    6,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    7,
    9,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    0,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    5,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    7,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    8,
    10,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    0,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    5,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    8,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    9,
    11,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    0,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    6,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    9,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    10,
    12,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    0,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    6,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    9,
  ],
  Array [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    11,
  ],
]
`;

exports[`atLeastFloat 2 1`] = `
Array [
  10.1,
  15.3,
  17.9,
  19.2,
  19.85,
  20.175,
  20.3375,
  20.41875,
  20.459375,
  20.4796875,
  20.48984375,
  20.494921875,
  20.4974609375,
  20.49873046875,
  20.499365234375,
  20.4996826171875,
  20.49984130859375,
  20.499920654296876,
]
`;

exports[`atLeastFloat 2 2`] = `
Array [
  -5,
  -7.5,
  -8.75,
  -9.375,
  -9.6875,
  -9.84375,
  -9.921875,
  -9.9609375,
  -9.98046875,
  -9.990234375,
  -9.9951171875,
  -9.99755859375,
  -9.998779296875,
  -9.9993896484375,
  -9.99969482421875,
  -9.999847412109375,
  -9.999923706054688,
]
`;

exports[`atLeastFloat 2 3`] = `
Array [
  -0,
  -10,
  -15,
  -17.5,
  -18.75,
  -19.375,
  -19.6875,
  -19.84375,
  -19.921875,
  -19.9609375,
  -19.98046875,
  -19.990234375,
  -19.9951171875,
  -19.99755859375,
  -19.998779296875,
  -19.9993896484375,
  -19.99969482421875,
  -19.999847412109375,
  -19.999923706054688,
]
`;

exports[`atLeastFloat 2 4`] = `
Array [
  0,
  2.5,
  3.75,
  4.375,
  4.6875,
  4.84375,
  4.921875,
  4.9609375,
  4.98046875,
  4.990234375,
  4.9951171875,
  4.99755859375,
  4.998779296875,
  4.9993896484375,
  4.99969482421875,
  4.999847412109375,
  4.9999237060546875,
]
`;

exports[`atLeastInteger 2 1`] = `
Array [
  5,
  13,
  17,
  19,
]
`;

exports[`atLeastInteger 2 2`] = `
Array [
  1,
  6,
  8,
  9,
]
`;

exports[`atLeastInteger 2 3`] = `
Array [
  9,
]
`;

exports[`atLeastInteger 2 4`] = `
Array [
  -5,
  -8,
  -9,
]
`;

exports[`atLeastInteger 2 5`] = `
Array [
  -0,
  -10,
  -15,
  -18,
  -19,
]
`;

exports[`char/atLeastChar/character 1`] = `
Array [
  " ",
  "=",
  "L",
  "S",
  "W",
  "Y",
]
`;

exports[`float 1`] = `
Array [
  0,
  10.25,
  15.375,
  17.9375,
  19.21875,
  19.859375,
  20.1796875,
  20.33984375,
  20.419921875,
  20.4599609375,
  20.47998046875,
  20.489990234375,
  20.4949951171875,
  20.49749755859375,
  20.498748779296875,
  20.499374389648438,
  20.49968719482422,
  20.49984359741211,
  20.499921798706055,
]
`;

exports[`float 2`] = `
Array [
  -0,
  -5.05,
  -7.574999999999999,
  -8.837499999999999,
  -9.46875,
  -9.784375,
  -9.9421875,
  -10.021093749999999,
  -10.060546875,
  -10.0802734375,
  -10.09013671875,
  -10.095068359374999,
  -10.0975341796875,
  -10.09876708984375,
  -10.099383544921874,
  -10.099691772460936,
  -10.099845886230469,
  -10.099922943115235,
]
`;

exports[`integer 1`] = `
Array [
  0,
  10,
  15,
  18,
  19,
]
`;

exports[`integer 2`] = `
Array [
  -0,
  -5,
  -8,
  -9,
]
`;

exports[`integer 3`] = `
Array [
  0,
  2500,
  3750,
  4375,
  4688,
  4844,
  4922,
  4961,
  4981,
  4991,
  4996,
  4998,
  4999,
]
`;

exports[`string 1`] = `
Array [
  "",
  "567890qq",
  "gabe1234",
  "1234567890qq",
  "567890qqgabe",
  "90qq1234gabe",
  "56781234gabe",
  "be1234567890qq",
  "1234567890qqga",
  "34567890qqbega",
  "567890qq12bega",
  "7890qq3412bega",
  "90qq563412bega",
  "qq78563412bega",
  "9078563412bega",
  "abe1234567890qq",
  "be1234567890qqg",
  "e1234567890qqag",
  "1234567890qqbag",
  "234567890qqebag",
  "34567890qq1ebag",
  "4567890qq21ebag",
  "567890qq321ebag",
  "67890qq4321ebag",
  "7890qq54321ebag",
  "890qq654321ebag",
  "90qq7654321ebag",
  "0qq87654321ebag",
  "qq987654321ebag",
  "q0987654321ebag",
  "q0987654321ebag",
  " abe1234567890qq",
  "Dabe1234567890qq",
  "Vabe1234567890qq",
  "_abe1234567890qq",
  "cabe1234567890qq",
  "eabe1234567890qq",
  "fabe1234567890qq",
  "g be1234567890qq",
  "gAbe1234567890qq",
  "gQbe1234567890qq",
  "gYbe1234567890qq",
  "g]be1234567890qq",
  "g_be1234567890qq",
  "g\`be1234567890qq",
  "ga e1234567890qq",
  "gaAe1234567890qq",
  "gaRe1234567890qq",
  "gaZe1234567890qq",
  "ga^e1234567890qq",
  "ga\`e1234567890qq",
  "gaae1234567890qq",
  "gab 1234567890qq",
  "gabC1234567890qq",
  "gabT1234567890qq",
  "gab]1234567890qq",
  "gaba1234567890qq",
  "gabc1234567890qq",
  "gabd1234567890qq",
  "gabe 234567890qq",
  "gabe)234567890qq",
  "gabe-234567890qq",
  "gabe/234567890qq",
  "gabe0234567890qq",
  "gabe1 34567890qq",
  "gabe1)34567890qq",
  "gabe1.34567890qq",
  "gabe1034567890qq",
  "gabe1134567890qq",
  "gabe12 4567890qq",
  "gabe12*4567890qq",
  "gabe12/4567890qq",
  "gabe1214567890qq",
  "gabe1224567890qq",
  "gabe123 567890qq",
  "gabe123*567890qq",
  "gabe123/567890qq",
  "gabe1232567890qq",
  "gabe1233567890qq",
  "gabe1234 67890qq",
  "gabe1234+67890qq",
  "gabe1234067890qq",
  "gabe1234367890qq",
  "gabe1234467890qq",
  "gabe12345 7890qq",
  "gabe12345+7890qq",
  "gabe1234517890qq",
  "gabe1234547890qq",
  "gabe1234557890qq",
  "gabe123456 890qq",
  "gabe123456,890qq",
  "gabe1234562890qq",
  "gabe1234565890qq",
  "gabe1234566890qq",
  "gabe1234567 90qq",
  "gabe1234567,90qq",
  "gabe1234567290qq",
  "gabe1234567590qq",
  "gabe1234567790qq",
  "gabe12345678 0qq",
  "gabe12345678-0qq",
  "gabe1234567830qq",
  "gabe1234567860qq",
  "gabe1234567880qq",
  "gabe123456789 qq",
  "gabe123456789(qq",
  "gabe123456789,qq",
  "gabe123456789.qq",
  "gabe123456789/qq",
  "gabe1234567890 q",
  "gabe1234567890Iq",
  "gabe1234567890]q",
  "gabe1234567890gq",
  "gabe1234567890lq",
  "gabe1234567890oq",
  "gabe1234567890pq",
  "gabe1234567890q ",
  "gabe1234567890qI",
  "gabe1234567890q]",
  "gabe1234567890qg",
  "gabe1234567890ql",
  "gabe1234567890qo",
  "gabe1234567890qp",
]
`;
